README.md
# koop-opendata

> An ArcGIS Open Data Provider for Koop

[![npm version][npm-img]][npm-url]

[npm-img]: https://img.shields.io/npm/v/koop-opendata.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/koop-opendata

`koop-opendata` allows you to take any search on ArcGIS Open Data and turn it into a Feature Service, GeoJSON, Shapefile or CSV.

## Install

To install or use this provider you will first need a working installation of Koop. We recommend the [`koop-sample-app`](https://github.com/koopjs/koop-sample-app) application template as an easy way to get started.

Add `koop-opendata` as a dependency to the `package.json` of your Node.js project.

```
npm install koop-opendata --save
```

## Usage

`koop-opendata` needs to be registered as a provider in your Koop app in order to work.

```js
var openData = require('koop-opendata')
koop.register(opendata)
```

After that you need to create an `openData:services` table in your spatial database.

```sql
CREATE TABLE "openData:services"
(
  id character varying(100),
  host character varying(100)
)
WITH (
  OIDS=FALSE
);
ALTER TABLE "openData:services"
  OWNER TO username;
```

Once that's done you can restart your server and the Open Data routes will be available.

If you're using the `koop-sample-app` template, you can start the server like this:

```
node server.js
```

### Register a site that you want to search against

- Example opendata.dc.gov
``` bash
curl -XPOST 'http://koop.com/openData' -d 'host=http://opendata.dc.gov&id=dc'
```
- Example opendata.arcgis.com
``` bash
curl -XPOST 'http://koop.com/openData' -d 'host=http://opendata.arcgis.com&id=umbrella'
```

### Start searching

- Example: Zoning datasets on opendata.dc.gov
```bash
curl -XGET 'http://koop.dc.esri.com/openData/dc/q/zoning'
```

- Example: Zoning datasets in the United States on opendata.arcgis.com
```bash
curl -XGET 'http://koop.dc.esri.com/openData/umbrella/q/zoning/place/united+states'
```

- Example: Land use datasets that have the keyword: zoning
```bash
curl -XGET 'http://koop.dc.esri.com/openData/umbrella/q/land+use/keyword/zoning'
```

- Example: Land use datasets, with the keyword zoning, sorted by date updated, in a bounding box
```bash
curl -XGET 'http://koop.dc.esri.com/openData/umbrella/keyword/zoning/sort_by/updated_at/bbox/135%2C1.014%2C-135%2C72.277'
```

* Note: the Koop URL is only for example purposes
* Not used to seeing curl commands? For the -XGET requests simply take the URL and paste it into your browser

### Search parameters

You can use all the search parameters together or choose to use none at all. Just place the parameter name before the one you want to use.

Examples:
- `/q/water`
- `/keyword/zoning`

Example of chained parameters:
- `/q/water/keyword/zoning`

```json
{
	"q": "a simple query string to search against",
	"keyword": "a keyword tag that must appear in any result",
	"bbox": "a bounding box to restrict results",
	"sort_by": "which way the results should be sorted [relevance, name, updated_at, created_at]",
	"place": "a place to restrict the results to. this place will be geocoded and used as a bounding box"
}
```

### Available formats

koop-opendata supports all the same formats as [Koop](http://github.com/esri/koop)

- KML -> simply append `.kml` to the request
- CSV -> append `.csv`
- Shapefile -> append `.zip`
- Feature Server -> append `/FeatureServer/0`
- GeoJSON -> append `.geojson`

## License

Copyright 2015 Esri

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

> http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

A copy of the license is available in the repository's [license.txt](license.txt) file.

[](Esri Tags: ArcGIS Web Mapping GeoJson FeatureServices)
[](Esri Language: JavaScript)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.4] - 2016-07-1
### Changed
* Using a different key struction for db and exports

## [1.1.3] - 2016-07-01
### Changed
* Bumping the version because previous publish was wrong

## 1.1.2 - 2015-07-01 [YANKED]
### Fixed
* Files are being dropped correctly

## [1.1.1] - 2015-07-01
### Changed
* Data is not returned until all the results have been processed

### Fixed
* Downloads contain all the results

## [1.1.0] - 2015-06-30
### Changed
* Added group property to return names of ArcGIS Online groups

## 1.0.1 - 2015-06-15
### Changed
* File exports are saved to a sub-directory

## 1.0.0 - 2015-06-15 [YANKED]
* initial release

[1.1.4]: https://github.com/koopjs/koop-opendata/compare/v1.1.3...v1.1.4
[1.1.3]: https://github.com/koopjs/koop-opendata/compare/v1.1.1...v1.1.3
[1.1.1]: https://github.com/koopjs/koop-opendata/compare/v1.1.1...v1.1.0
[1.1.0]: https://github.com/koopjs/koop-opendata/compare/v1.1.0...v1.0.1


test/model-test.js
/* global before, describe, it */

var should = require('should')
var config = require('config')
var kooplib = require('koop-server/lib')
var Sample = require('../models/Sample.js')
var model

before(function (done) {
  kooplib.Cache.db = kooplib.PostGIS.connect(config.db.postgis.conn)
  model = new Sample(kooplib)
  done()
})

describe('Sample Model', function () {
  describe('when getting data', function () {
    it('should find and return geojson', function (done) {
      model.find(1, {}, function (err, data) {
        // there should not be any errors
        should.not.exist(err)
        // should always return the data as geojson
        should.exist(data)
        // data should be an array (support multi layer responses)
        data.length.should.equal(1)
        // make sure we have a feature collection
        data[0].type.should.equal('FeatureCollection')
        done()
      })
    })

  })

})


test/controller-test.js
/* global after, before, describe, it */

var should = require('should')
var sinon = require('sinon')
var config = require('config')
var request = require('supertest')

// we require Koop so we can fake having an actual server running
var koop = require('koop-server')(config)

// we need koop/lib so we can have access to shared code not exposed directly off the koop object
var kooplib = require('koop-server/lib')

var sample

before(function (done) {
  // pull in the provider module
  var provider = require('../index.js')

  // create the model
  sample = new provider.model(kooplib) // eslint-disable-line

  // pass the model to the controller
  var controller = new provider.controller(sample) // eslint-disable-line

  // bind the default routes so we can test that those work
  koop._bindDefaultRoutes(provider.name, provider.pattern, controller)

  // bind the routes into Koop
  koop._bindRoutes(provider.routes, controller)
  done()
})

after(function (done) {
  done()
})

describe('Sample Controller', function () {
  describe('get', function () {
    before(function (done) {
      // we stub the find method so we dont actually try to call it
      // we're not testing the model here, just that the controller should call the model
      sinon.stub(sample, 'find', function (id, options, callback) {
        callback(null, [{
          type: 'FeatureCollection',
          features: [{ properties: {}, coordinates: {}, type: 'Feature' }]
        }])
      })

      done()
    })

    after(function (done) {
      // restore the stubbed methods so we can use them later if we need to
      sample.find.restore()
      done()
    })

    it('/sample/1 should call find', function (done) {
      request(koop)
        .get('/sample/1')
        .end(function (err, res) {
          should.not.exist(err)
          res.status.should.equal(200)
          // sample.find.called.should.equal(true)
          done()
        })
    })
  })

  describe('index', function () {
    it('/sample should return 200', function (done) {
      request(koop)
        .get('/sample')
        .end(function (err, res) {
          should.not.exist(err)
          res.status.should.equal(200)
          done()
        })
    })
  })

  describe('preview', function () {
    it('/sample/1/preview should return 200', function (done) {
      request(koop)
        .get('/sample/1/preview')
        .end(function (err, res) {
          should.not.exist(err)
          res.status.should.equal(200)
          done()
        })
    })
  })

  describe('FeatureServer', function () {
    it('/sample/1/FeatureServer should return 200', function (done) {
      request(koop)
        .get('/sample/1/FeatureServer')
        .end(function (err, res) {
          should.not.exist(err)
          res.status.should.equal(200)
          done()
        })
    })
  })

})


models/composer.js
// used to send http requests
var request = require('request')
// used to manage control flow
var async = require('async')
// used to create hashes that fingerprint a given request
var hash = require('object-hash')
var _ = require('underscore')
var bboxPolygon = require('turf-bbox-polygon')

var Composer = function (koop) {
  var composer = koop.BaseModel(koop)
  var type = 'composer'
  var token_expiration
  var locations = {}
  var token

  // adds a service to the koop.Cache.db
  // needs a host, generates an id
  composer.register = function (id, host, callback) {
    var type = 'composer:services'
    koop.Cache.db.serviceCount(type, function (error, count) {
      if (error) {
        return callback(error, null)
      }
      id = id || count++
      koop.Cache.db.serviceRegister(type, {'id': id, 'host': host}, function (err, success) {
        callback(err, id)
      })
    })
  }

  composer.remove = function (id, callback) {
    koop.Cache.db.serviceRemove('composer:services', parseInt(id, 0) || id, callback)
  }

  // get service by id, no id == return all
  composer.find = function (id, callback) {
    koop.Cache.db.serviceGet('composer:services', parseInt(id, 0) || id, function (err, res) {
      if (err) {
        callback('No datastores have been registered with this provider yet. Try POSTing {"host":"url", "id":"yourId"} to /composer', null)
      } else {
        callback(null, res)
      }
    })
  }

  composer.getResource = function (host, hostId, params, options, cb) {
    // delete these two keys or else we get inconsistent hash keys depending on the request
    params.layer = 0
    var key = hostId + hash.MD5(_.omit(params, 'method'))
    // check the cache for data with this type & id
    // if no prior requests exist then trigger this waterfall
    koop.Cache.get(type, key, options, function (err, entry) {
      if (err) {
        async.waterfall([
          function (callback) {
            if (!token || Date.now() >= token_expiration) {
              composer.getToken(function (err, res) {
                if (err) {
                  koop.log.error(err)
                } else {
                  token = res.body.access_token
                  callback(null)
                }
              })
            } else {
              callback(null)
            }
          },
          function (callback) {
            if (params.place) {
              composer.getBbox(params.place, function (bbox) {
                params.bbox = bbox
                callback(null)
              })
            } else {
              callback(null)
            }
          },
          function (callback) {
            var query = composer.buildOptions(params)
            callback(null, query)
          },
          function (query, callback) {
            composer.getDatasets(host, query, function (err, json) {
              if (err) {
                koop.log.error(err)
              } else {
                var count = json.metadata.stats.count
                if (count > 100) {
                  var pages = composer.buildPages(host, query, count)
                  composer.createQueue(key, pages)
                }
                callback(null, json)
              }
            })
          },
          function (json, callback) {
            callback(null, composer.translate(json))
          }
        ], function (err, geojson) {
          if (err) {
            koop.log.error(err)
          } else {
            composer.insert(key, geojson, cb)
          }
        })
      } else {
        cb(null, entry)
      }
    })
  }

  composer.insert = function (key, geojson, callback) {
    // take translated geojson and huck it into Koop
    koop.Cache.insert(type, key, geojson, 0, function (err, success) {
      if (err) {
        callback(err)
      } else {
        callback(null, success)
      }
    })
  }

  composer.insertPartial = function (key, geojson, callback) {
    koop.Cache.insertPartial(type, key, geojson, 0, function (err, success) {
      if (err) {
        callback(err)
      } else {
        callback(null, success)
      }
    })
  }

  composer.getDatasets = function (host, query, callback) {
    // simple wrapper around requests to the desired API
    var url = host + '/datasets.json?' + query
    request.get(url, function (err, res, body) {
      callback(err, JSON.parse(body))
    })
  }

  composer.getToken = function (callback) {
    request.post({
      url: 'https://www.arcgis.com/sharing/rest/oauth2/token/',
      json: true,
      form: {
        'f': 'json',
        'client_id': koop.config.esri.id,
        'client_secret': koop.config.esri.secret,
        'grant_type': 'client_credentials',
        'expiration': '60'
      }
    }, function (err, res) {
      callback(err, res)
    })
  }

  composer.getBbox = function (place, callback) {
    // takes in a location string and returns a bbox in the format composer understands
    var bbox
    if (locations[place]) {
      bbox = locations.place
      callback(null, bbox)
    } else {
      var root = 'http://geocode.arcgis.com'
      var gc_request = '/arcgis/rest/services/World/GeocodeServer/find?f=json&forStorage=true&maxlocations=1&outSR=4326'
      gc_request = root + gc_request + '&text=' + encodeURI(place) + '&token=' + token
      request.get(gc_request, function (err, res) {
        // handle the geocoder result
        if (err) {
          callback(err)
        } else {
          var response = JSON.parse(res.body)
          var extent = response.locations[0].extent
          bbox = extent.xmin + ',' + extent.ymin + ',' + extent.xmax + ',' + extent.ymax
          locations.place = bbox
          callback(null, bbox)
        }
      })
    }
  }

  composer.buildPages = function (host, query, count) {
    var pageCount = Math.ceil(count / 100) - 1
    var pages = []
    for (var p = 0; p < pageCount; p++) {
      var page = host + query + '&page=' + p + 2
      pages.push(page)
    }
    return pages
  }

  composer.createQueue = function (key, pages) {
    var pageQueue = async.queue(function (page, callback) {
      composer.getDatasets(page, function (err, json) {
        if (err) {
          return callback(err)
        }
        var geojson = composer.translate(json)
        composer.insertPartial(key, geojson, function (err, success) {
          if (err) {
            return callback(err)
          }
          callback(null, success)
        })
      })
    }, 4)

    pageQueue.drain = function () {
      koop.log.info('Finished paging: ' + key)
    }

    pageQueue.push(pages, function () {
      koop.log.info('Sucessfully processed a page of ' + key)
    })

    return koop.log.info('Paging kicked off for ' + key)
  }

  composer.buildOptions = function (params) {
    // create a a default set of parameters for the API call
    // fill in passed in parameters where available
    var options = {
      q: '*' || params.query,
      sort_by: 'relevance' || params.sort_by,
      per_page: 100
    }
    if (params.keyword) {
      options.keyword = params.keyword
    }
    if (params.bbox) {
      options.bbox = params.bbox
      options.filter_by_extent = true
    }

    // concatenate all the parameters into one big string
    var parameters = ''
    for (var key in options) {
      parameters = parameters + key + '=' + options[key] + '&'
    }
    return parameters.slice(0, -1)
  }

  composer.translate = function (json) {
    // translate the composer API response into geojson
    // create the shell that will hold all the properties
    var geojson = {
      type: 'FeatureCollection',
      features: []
    }
    json.data.forEach(function (dataset) {
    // loop through each dataset returned from the API call and push it into the geojson shell
      var coordinates = bboxPolygon(dataset.extent[0].concat(dataset.extent[1]))
      geojson.features.push({
        type: 'Feature',
        geometry: {
          type: 'Point',
          coordinates: coordinates
        },
        properties: {
          name: dataset.name,
          owner: dataset.owner,
          updatedAt: dataset.updated_at,
          quality: dataset.quality,
          coverage: dataset.coverage,
          description: dataset.description,
          tags: dataset.tags.join(', '),
          server: dataset.url,
          serverVersion: dataset.current_version,
          license: dataset.license,
          records: dataset.record_count,
          thumbnail: dataset.thumbnail_url,
          arcGisUrl: dataset.arcgis_online_item_url
        }
      })
    })
    return geojson
  }

  composer.drop = function (key, options, callback) {
  // drops the item from the cache
    var dir = ['composer', key, 0].join(':')
    koop.Cache.remove('composer', key, options, function (err, res) {
      koop.files.removeDir('files/' + dir, function (err, res) {
        koop.files.removeDir('tiles/' + dir, function (err, res) {
          koop.files.removeDir('thumbs/' + dir, function (err, res) {
            callback(err, true)
          })
        })
      })
    })
  }

  return composer

}

module.exports = Composer


controller/index.js
var crypto = require('crypto')
var Path = require('path')

// a function that is given an instance of Koop at init
var Controller = function (OpenData, BaseController) {
  var controller = BaseController()

  // register a OpenData instance
  controller.register = function (req, res) {
    if (!req.body.host) {
      res.send('Must provide a host to register:', 500)
    } else {
      OpenData.register(req.body.id, req.body.host, function (err, id) {
        if (err) {
          res.send(err, 500)
        } else {
          res.json({ 'serviceId': id })
        }
      })
    }
  }

  controller.list = function (req, res) {
    OpenData.find(null, function (err, data) {
      if (err) {
        res.send(err, 500)
      } else {
        res.json(data)
      }
    })
  }

  controller.find = function (req, res) {
    OpenData.find(req.params.id, function (err, data) {
      if (err) {
        res.send(err, 404)
      } else {
        res.json(data)
      }
    })
  }

  // drops the cache for an item
  controller.drop = function (req, res) {
    OpenData.find(req.params.id, function (err, data) {
      if (err) {
        res.send(err, 500)
      } else {
        // Get the item
        OpenData.drop(data.id, req.params, req.query, function (error, itemJson) {
          if (error) {
            res.send(error, 500)
          } else {
            res.json(itemJson)
          }
        })
      }
    })
  }

  controller.findResource = function (req, res) {
    OpenData.find(req.params.id, function (err, data) {
      if (err) {
        res.status(500).send(err)
      } else {
        // Get the item
        OpenData.getResource(data.host, data.id, req.params, req.query, function (error, itemJson) {
          if (error) {
            return res.status(500).send(error)
            // if there's no filter and we don't have all the search results just send back processing
          } else if (!req.query.length && itemJson[0].features.length < itemJson[0].totalCount) {
            return res.status(200).json([{processing: true}])
          } else if (req.params.format) {
            // change geojson to json
            var format = req.params.format.replace('geojson', 'json')
            delete req.params.format
            // build the file key as an MD5 hash that's a join on the params and look for the file
            var stringedParams = JSON.stringify(req.params)
            var toHash = stringedParams + JSON.stringify(req.query)
            var key = crypto.createHash('md5').update(toHash).digest('hex')
            var dir = 'OpenData' + '/' + req.params.id + '/' + crypto.createHash('md5').update(stringedParams).digest('hex')
            var filePath = ['files', dir, key].join('/')
            var fileName = key + '.' + req.params.format
            OpenData.files.exists(filePath, fileName, function (exists, path) {
              if (exists) {
                if (path.substr(0, 4) === 'http') {
                  res.redirect(path)
                } else {
                  res.sendFile(path)
                }
              } else {
                OpenData.exportToFormat(format, dir, key, itemJson[0], {rootDir: OpenData.files.localDir}, function (err, file) {
                  if (err) return res.status(500).send(err)
                  res.status(200).sendFile(Path.resolve(process.cwd(), file.file))
                })
              }
            })
          } else {
            var geojson = itemJson
            res.status(200).json(geojson[0])
          }
        })
      }
    })
  }

  controller.del = function (req, res) {
    if (!req.params.id) {
      res.send('Must specify a service id', 500)
    } else {
      OpenData.remove(req.params.id, function (err, data) {
        if (err) {
          res.status(500).send(err)
        } else {
          res.json(data)
        }
      })
    }
  }
  // shared dispath for feature service responses
  controller.featureserver = function (req, res) {
    var callback = req.query.callback
    delete req.query.callback
    for (var k in req.body) {
      req.query[k] = req.body[k]
    }
    OpenData.find(req.params.id, function (err, data) {
      if (err) {
        res.status(500).send(err)
      } else {
        var host = data.host
        // if this is a count request then go straight to the db
        if (req.query.returnCountOnly) {
          controller.featureserviceCount(req, res, host)
        } else {
          // else send this down for further processing
          controller.featureservice(req, res, host, callback)
        }
      }
    })
  }

  controller.featureserviceCount = function (req, res, host) {
    // first check if the dataset is new, in the cache, or processing
    // ask for a single feature becasue we just want to know if the data is there
    req.query.limit = 1
    OpenData.getResource(host, req.params.id, req.params.item, req.query, function (err, geojson) {
      if (err) {
        res.status(500).send(err)
      } else if (geojson[0] && geojson[0].status === 'processing') {
        res.status(202).json(geojson)
      } else {
        // it's not processing so send for the count
        OpenData.getCount(['OpenData', req.params.item, (req.query.layer || 0)].join(':'), req.query, function (err, count) {
          if (err) {
            console.log('Could not get feature count', req.params.item)
            res.status(500).send(err)
          } else {
            var response = {count: count}
            res.status(200).json(response)
          }
        })
      }
    })
  }

  controller.featureservice = function (req, res, host, callback) {
    var err
    req.query.limit = req.query.limit || req.query.resultRecordCount || 1000000000
    req.query.offset = req.query.resultOffset || null
    // Get the item
    OpenData.getResource(host, req.params.id, req.params, req.query, function (error, geojson) {
      if (error) {
        res.status(500).send(error)
      } else if (geojson[0] && geojson[0].status === 'processing') {
        res.status(202).json(geojson)
      } else {
        // pass to the shared logic for FeatureService routing
        delete req.query.geometry
        delete req.query.where
        controller.processFeatureServer(req, res, err, geojson, callback)
      }
    })
  }

  controller.preview = function (req, res) {
    res.render(__dirname + '/../views/demo', {
      locals: {
        host: req.params.id,
        item: req.params.item
      }
    })
  }

  return controller
}

module.exports = Controller


models/OpenData.js
// used to send http requests
var crypto = require('crypto')
var request = require('request')
// used to manage control flow
var async = require('async')
// used to create hashes that fingerprints a given request
var hash = require('object-hash')
var _ = require('underscore')
var bboxPolygon = require('turf-bbox-polygon')

var OpenData = function (koop) {
  var openData = koop.BaseModel(koop)
  var type = 'OpenData'
  var tokenExpiration
  var locations = {}
  var token

  // adds a service to the koop.Cache.db
  // needs a host, generates an id
  openData.register = function (id, host, callback) {
    var type = 'openData:services'
    koop.Cache.db.serviceCount(type, function (error, count) {
      if (error) {
        return callback(error, null)
      }
      id = id || count++
      koop.Cache.db.serviceRegister(type, {'id': id, 'host': host}, function (err, success) {
        callback(err, id)
      })
    })
  }

  openData.remove = function (id, callback) {
    koop.Cache.db.serviceRemove('openData:services', parseInt(id, 0) || id, callback)
  }

  // get service by id, no id == return all
  openData.find = function (id, callback) {
    koop.Cache.db.serviceGet('openData:services', parseInt(id, 0) || id, function (err, res) {
      if (err) {
        callback('No datastores have been registered with this provider yet. Try POSTing {"host":"url", "id":"yourId"} to /openData', null)
      } else {
        callback(null, res)
      }
    })
  }

  openData.getResource = function (host, hostId, params, options, callback) {
    params.layer = 0
    // delete the mehod param or else we get inconsistent hash keys depending on the request
    var key = hostId + '_' + hash.MD5(_.omit(params, 'method'))
    // check the cache for data with this type & id
    koop.Cache.get(type, key, options || {layer: 0}, function (err, entry) {
      if (err) {
        openData.buildQuery(params, function (err, query) {
          if (err) return koop.log.error(err)
          openData.search(key, host, query, function (err, geojson) {
            if (err) return callback(err)
            callback(null, [geojson])
          })
        })
      } else {
        callback(null, entry)
      }
    })
  }

  openData.search = function (key, host, query, callback) {
    var queryString = host + '/datasets?' + query
    // get the first page and figure out the count
    openData.getPage(host, query, function (err, results) {
      if (err) return callback(err)
      var count = results.metadata.stats.total_count
      koop.log.info(queryString + ' has ' + count + 'results')
      openData.translate(host, results, function (err, geojson) {
        if (err) return callback(err)
        geojson.name = queryString
        geojson.totalCount = count
        // insert the first page then create a queue to handle the rest if we need to
        openData.insert(key, geojson, function (err, success) {
          if (err) return callback(err)
          // return results as soon as there is anything cached
          // but let the client know data is still being retreived.
          if (count > 100) geojson.processing = true
          callback(null, geojson)
          var pages = openData.buildPages(query, count)
          // if more than 1 page, gather in the background
          if (count > 100) {
            openData.createQueue(key, host, pages, function (err, info) {
              if (err) return koop.log.error(err)
              koop.log.info(info)
            })
          }
        })
      })
    })
  }

  openData.insert = function (key, geojson, callback) {
    // take translated geojson and huck it into Koop
    koop.Cache.insert(type, key, geojson, 0, function (err, success) {
      if (err) {
        callback(err)
      } else {
        callback(null, success)
      }
    })
  }

  openData.insertPartial = function (key, geojson, callback) {
    koop.Cache.insertPartial(type, key, geojson, 0, function (err, success) {
      if (err) {
        callback(err)
      } else {
        callback(null, success)
      }
    })
  }

  openData.getPage = function (host, query, callback) {
    // simple wrapper around requests to ArcGIS Open Data API
    var url = host + '/datasets.json?' + query
    var options = {
      url: url,
      gzip: true
    }
    request.get(options, function (err, res, body) {
      if (err) return callback(err)
      try {
        callback(null, JSON.parse(body))
      } catch (e) {
        callback(url + '::' + e)
      }
    })
  }

  openData.buildPages = function (query, count) {
    var pageCount = Math.ceil(count / 100) + 1
    var pages = []
    for (var p = 2; p < pageCount; p++) {
      var page = query + '&page=' + p
      pages.push(page)
    }
    return pages
  }

  openData.createQueue = function (key, host, pages, callback) {
    var pageQueue = async.queue(function (page, cb) {
      openData.getPage(host, page, function (err, results) {
        if (err) return cb(err)
        openData.translate(host, results, function (err, geojson) {
          if (err) return cb(err)
          openData.insertPartial(key, geojson, function (err, success) {
            if (err) return cb(err)
            cb()
          })
        })
      })
    }, 4)

    pageQueue.drain = function () {
      callback(null, 'Finished paging: ' + key)
    }
    pageQueue.push(pages, function (err) {
      if (err) return callback(err)
      koop.log.info('Sucessfully processed a page of ' + key)
    })

    callback(null, 'Paging kicked off for ' + key)
  }

  openData.buildQuery = function (params, callback) {
    // create a a default set of parameters for the API call
    // fill in passed in parameters where available
    var options = {
      q: params.q || '*',
      sort_by: params.sort_by || 'relevance',
      per_page: 100
    }
    if (params.keyword) {
      options.keyword = params.keyword
    }
    // accept bbox parameter or geocode a place parameter
    openData.getBbox(params, function (err, bbox) {
      if (err) return callback(err)
      if (bbox) {
        options.bbox = bbox
        options.filter_by_extent = true
      }
      // concatenate all the parameters into one query string
      var query = ''
      for (var key in options) {
        query += key + '=' + options[key] + '&'
      }
      callback(null, query.slice(0, -1))
    })
  }

  openData.getBbox = function (params, callback) {
    var bbox
    // there is no bbox or place parameter
    if (!params.bbox && !params.place) {
      callback(null, null)
    // already have a bbox don't need to geocode
    } else if (params.bbox) {
      callback(null, bbox)
    // already geocoded this place parameter
    } else if (locations[params.place]) {
      callback(null, locations[params.place])
    // new place parameter, get the bounding box
    } else {
      openData.getToken(function (err, token) {
        if (err) {
          koop.log.error('Error getting Esri Geocoding token, disregarding place parameter')
          return callback(null, null)
        }
        var root = 'http://geocode.arcgis.com'
        var gc_request = '/arcgis/rest/services/World/GeocodeServer/find?f=json&forStorage=true&maxlocations=1&outSR=4326'
        gc_request = root + gc_request + '&text=' + encodeURI(params.place) + '&token=' + token
        request.get(gc_request, function (err, res) {
          if (err) return callback(err)
          var response = JSON.parse(res.body)
          var extent = response.locations[0].extent
          bbox = extent.xmin + ',' + extent.ymin + ',' + extent.xmax + ',' + extent.ymax
          // save the result to an in-memory object to avoid repeat requests
          locations[params.place] = bbox
          callback(null, bbox)
        })
      })
    }
  }

  openData.getToken = function (callback) {
    if (token && new Date() < tokenExpiration) {
      callback(null, token)
    } else {
      request.post({
        url: 'https://www.arcgis.com/sharing/rest/oauth2/token/',
        json: true,
        form: {
          'f': 'json',
          'client_id': koop.config.esri.id,
          'client_secret': koop.config.esri.secret,
          'grant_type': 'client_credentials',
          'expiration': '60'
        }
      }, function (err, res) {
        if (err) return callback(err)
        token = res.body.access_token
        tokenExpiration = new Date(res.body.expires + (1000 * 60 * 60))
        callback(null, token)
      })
    }
  }

  openData.translate = function (host, results, callback) {
    // translate the openData API response into geojson
    // create the shell that will hold all the properties
    var geojson = {
      type: 'FeatureCollection',
      features: []
    }
    results.data.forEach(function (dataset) {
    // loop through each dataset returned from the API call and push it into the geojson shell
      // console.log(dataset)
      try {
        var bbox = dataset.extent.coordinates
        if (bbox) {
          var coordinates = bboxPolygon(bbox[0].concat(bbox[1])).geometry.coordinates
          geojson.features.push({
            type: 'Feature',
            geometry: {
              type: 'Polygon',
              coordinates: coordinates
            },
            properties: {
              name: dataset.name,
              owner: dataset.owner,
              description: dataset.description,
              updatedAt: dataset.updated_at,
              quality: dataset.quality,
              coverage: dataset.coverage,
              tags: dataset.tags.join(', '),
              openDataUrl: host + '/datasets/' + dataset.id,
              server: dataset.url,
              serverVersion: dataset.current_version,
              license: dataset.license,
              records: dataset.record_count,
              thumbnail: dataset.thumbnail_url,
              arcGisUrl: dataset.arcgis_online_item_url,
              group: dataset.main_group_title
            }
          })
        }
      } catch (e) {
        // todo should I just continue here?
        callback(e)
      }
    })
    callback(null, geojson)
  }

  openData.exportToFormat = function (format, dir, key, geojson, options, callback) {
    koop.Exporter.exportToFormat(format, dir, key, geojson, options, callback)
  }

  openData.drop = function (hostId, params, options, callback) {
  // drops the item from the cache
    delete params.method
    params.layer = 0
    var stringedParams = JSON.stringify(params)
    var dir = 'OpenData' + '/' + params.id + '/' + crypto.createHash('md5').update(stringedParams).digest('hex')
    var key = hostId + ':' + hash.MD5(_.omit(params, 'method'))
    koop.Cache.remove(type, key, {}, function (err, res) {
      if (err) return callback(err)
      koop.files.removeDir('files/' + dir, function (err, res) {
        if (err) return callback(err)
        koop.files.removeDir('tiles/' + dir, function (err, res) {
          if (err) return callback(err)
          koop.files.removeDir('thumbs/' + dir, function (err, res) {
            if (err) return callback(err)
            callback(err, true)
          })
        })
      })
    })
  }

  openData.getCount = function (key, options, callback) {
    koop.Cache.getCount(key, options, callback)
  }

  return openData

}

module.exports = OpenData
